Re-do the backend of `@where`

pdeffebach opened the pull request:
Consistent with the new backend of `@orderby` in #191 , I am writing a new backend for `@where`

1. Always returns a `DataFrame`
2. Makes a transformation by group and then filters

Need to add tests and write docs still.

pdeffebach commented:
@bkamins, @nalimilan 

Just to confirm before moving forward with this, despite the problems with `@orderby` with a GroupedDataFrame, we *do* want `@where` to perform tranformations by group and then return a DataFrame, right? 

Unlike `@orderby`, this feature will be widely used. 

```
@where(df, :x .> quantile(:x, .9)
```

for example, keep the rows with the top 10% `:x` value in each group.

bkamins commented:
In my words `@where` would get whole columns on input and do "within group" filtering and return a `DataFrame`. Right?

Would it drop `missing`? I assume yes.

What would be the order of rows? I assume the same as in the source (so it would not order rows by group but just filter the original data frame).

We have not discussed it yet (and I have not thought on it much), but I understand that this is also what `where` in DataFrames.jl would do. Then we would document `where` as working in the same way as `select` and `transform` with the exception that it drops rows.

pdeffebach commented:
Yes, I think it would treat `missing` as `false`. 

Great. I'm glad you agree. I will move forward with this, then.

bkamins commented:
There is one detail here. `@where` will not work on subsetted `GroupedDataFrame` (and obviously it will ignore group order). I would be good to document both things.

pdeffebach commented:
> There is one detail here. `@where` will not work on subsetted `GroupedDataFrame`

Can you clarify what `subsetted` grouped data frame means?

bkamins commented:
you can subset a `GroupedDataFrame` like `gdf[[1,3,5]]` (where you keep only groups 1,3, and 5 in this example)

pdeffebach commented:
Ah okay. I will make sure to document that!

pdeffebach commented:
Ready for a review!

bkamins commented on src/DataFramesMeta.jl:
```diff
@@ -116,10 +116,19 @@ function fun_to_vec(kw::Expr; nolhs = false)
-            t = quote
-                DataFramesMeta.make_source_concrete($(source)) =>
-                ($(Expr(:tuple, values(membernames)...)) -> $body)
+            if gensym_names == false
```
maybe change order and just use `gensym_names` as a condition?

bkamins commented on src/DataFramesMeta.jl:
```diff
@@ -100,7 +100,7 @@ end
 # We don't create the "new name" pair because new names are given
 # by the table.
-function fun_to_vec(kw::Expr; nolhs = false)
+function fun_to_vec(kw::Expr; nolhs = false, gensym_names = false)
```
```suggestion
function fun_to_vec(kw::Expr; nolhs::Bool=false, gensym_names::Bool=false)
```

bkamins commented on src/DataFramesMeta.jl:
```diff
@@ -141,7 +150,7 @@ function fun_to_vec(kw::Expr; nolhs = false)
 end
 
-fun_to_vec(kw::QuoteNode; nolhs = false) = kw
+fun_to_vec(kw::QuoteNode; nolhs = false, gensym_names = false) = kw
```
same here

bkamins commented on src/DataFramesMeta.jl:
```diff
@@ -295,20 +304,40 @@ end
 
-and(x, y) = :($x .& $y)
+function df_to_bool(res)
+    if any(t -> !(t isa Union{Vector{<:Union{Missing, Bool}}, BitArray{1}}), eachcol(res))
```
why `AbstractVector` is not allowed? E.g. from a `SubDataFrame`? Is this tested?

pdeffebach replied:
Thanks for this. I added a method for `SubDataFrame` for both `@where` and `@orderby` because of this. (needs a new method because `select(subdf, ..., copycols = false)` fails.

bkamins replied:
it is expected to fail if `copycols=false` only column reusing and then `SubDataFrame` is produced (no transformations) are allowed.

bkamins commented on src/DataFramesMeta.jl:
```diff
@@ -295,20 +304,40 @@ end
 
-function where_helper(d, args...)
-    :($where($d, $(with_anonymous(reduce(and, args)))))
+function where(gd::GroupedDataFrame, @nospecialize(args...))
```
why is this defined?

pdeffebach replied:
I thought we decided that we wanted `where` to work on a GroupedDataFrame.

bkamins replied:
yes - I mixed it up with `orderby`

bkamins commented on src/DataFramesMeta.jl:
```diff
@@ -295,20 +304,40 @@ end
     @where(d, i...)
 
-Select row subsets in AbstractDataFrames or groups in GroupedDataFrames.
+Select row subsets in AbstractDataFrames and GroupedDataFrames.
```
only data frames

bkamins commented on src/DataFramesMeta.jl:
```diff
@@ -346,48 +386,43 @@ julia> @where(df, :x .> x, :y .== 3)
-│ 4   │ 15 │ 2 │
-│ 5   │ 16 │ 2 │
-│ 6   │ 20 │ 2 │
+julia> @where(g, :n .> mean(:n))
```
you still have examples with `gdf`.

pdeffebach commented:
Thanks! All your commends should be addressed.

pdeffebach commented:
Another note, 

```
@where(gd[[1:2]], args...)
```

fails at the stage of `DataFrames.select`, which I think is fine.

bkamins commented:
> at the stage of `DataFrames.select`

As you know this is disallowed and intended :) if `gdf` has more than 2 groups. 😄.

Do you think we want to go back to the discussion if `select` on `gdf[[1:2]]` should be allowed?

pdeffebach commented:
No, I am happy with the current scenario. Maybe people will request that feature in the future, but I'm fine with whatever DataFrames does.

pdeffebach commented:
Tests pass on 1.0, fail on nightly for an unrelated reason. 

We should probably start testing 1.0 and 1.6-dev rather than nightly.

nalimilan commented on src/DataFramesMeta.jl:
```diff
@@ -295,20 +299,46 @@ end
+and(x, y) = x .& y
+
+function df_to_bool(res)
+    if any(t -> !(t isa Union{AbstractVector{<:Union{Missing, Bool}}, BitArray{1}}), eachcol(res))
+        throw(ArgumentError("All arguments in @where must return a " *
+                            "AbstractVector{<:Union{Missing, Bool} or a BitArray{1}"))
```
`BitArray{1} <: AbstractVector{<:Union{Missing, Bool}` already:
```suggestion
    if any(t -> !(t isa AbstractVector{<:Union{Missing, Bool}}), eachcol(res))
        throw(ArgumentError("All arguments in @where must return an " *
                            "AbstractVector{<:Union{Missing, Bool}}"))
```

nalimilan commented on src/DataFramesMeta.jl:
```diff
@@ -295,20 +299,46 @@ end
+
+and(x, y) = x .& y
+
+function df_to_bool(res)
```
```suggestion
function df_to_bool(res::AbstractDataFrame)
```

nalimilan commented on src/DataFramesMeta.jl:
```diff
@@ -295,20 +299,46 @@ end
+                            "AbstractVector{<:Union{Missing, Bool} or a BitArray{1}"))
+    end
+
+    return reduce(and, eachcol(res)) .=== true
```
Why not this?
```suggestion
    return reduce((x, y) -> x .& y, eachcol(res)) .=== true
```

pdeffebach replied:
I am still getting a grip on how anonymous functions change compilation time. You are right that this doesn't matter.

nalimilan commented on src/DataFramesMeta.jl:
```diff
@@ -295,20 +299,46 @@ end
     @where(d, i...)
 
-Select row subsets in AbstractDataFrames or groups in GroupedDataFrames.
+Select row subsets in AbstractDataFrames and GroupedDataFrames.
```
```suggestion
Select row subsets in `AbstractDataFrame`s and `GroupedDataFrame`s.
```

nalimilan commented on src/DataFramesMeta.jl:
```diff
@@ -317,6 +347,17 @@ Select row subsets in AbstractDataFrames or groups in GroupedDataFrames.
 
 Multiple `i` expressions are "and-ed" together.
 
+If given a `GroupedDataFrame`, `@where` applies transformations by
+group, but does not store the result. `@where` returns a fresh
+`DataFrame` where the rows generated from the transformations are all
+`true`.
```
```suggestion
If given a `GroupedDataFrame`, `@where` applies transformations by
group, and returns a fresh `DataFrame` containing the rows
for which the generated values are all `true`.
```

nalimilan commented on test/grouping.jl:
```diff
@@ -291,4 +287,19 @@ end
+
+    gd = groupby(df, :g)
+
+    @test @where(gd, :i .== first(:i)).t == ["a", "c"]
+    @test @where(gd, cols(:i) .> mean(cols(:i)), :t .== "c").t == ["c"]
+    @test @where(gd, :c == :g).i == Int[]
```
As usual I'd recommend testing equality against a full data frame, to ensure everything is correct and not just one column.

nalimilan commented on src/DataFramesMeta.jl:
```diff
@@ -413,6 +449,11 @@ function orderby(x::GroupedDataFrame, @nospecialize(args...))
     throw(ArgumentError("@orderby with a GroupedDataFrame is reserved"))
 end
 
+function orderby(x::SubDataFrame, @nospecialize(args...))
+    t = DataFrames.select(x, args...)
+    x[sortperm(t), :]
+end
```
Why is this in this PR?

pdeffebach replied:
Because its a bug that Bogumil noticed in the implementation of `@where` that we hadn't caught in the implementation of `@orderby`. I can delete it if you want to keep things separate.

nalimilan commented on test/dataframes.jl:
```diff
@@ -286,11 +300,15 @@ end
     @test @orderby(df, :c).i == [3, 1, 2, 4, 5]
     @test @orderby(df, -:g).i == [4, 5, 1, 2, 3]
     @test @orderby(df, :t).i == [1, 2, 3, 4, 5]
+
+    @test @orderby(df, identity(:g), :g.^2).i == [1, 2, 3, 4, 5]
+
+    subdf = @view df[1:3, :]
+
+    @test @orderby(subdf, -:i).i == [3, 2, 1]
```
Same remark here.

pdeffebach commented:
Thanks for the review. I believe I have addressed all your comments. 

Sorry about tacking on the `@orderby` fix onto this PR. It seemed natural since it's the same implementation. I can split those changes into another PR if needed.

pdeffebach commented:
Thanks!
